plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtConnection.java

```diff
@@ -101,4 +113,33 @@ static String getHypervisorURI(String hypervisorType) {
 
         return "qemu:///system";
     }
+
+    /**
+     * Set up Libvirt event handling and polling. This is not specific to a connection object instance, but needs
+     * to be done prior to creating connections. See the Libvirt documentation for virEventRegisterDefaultImpl and
+     * virEventRunDefaultImpl or the libvirt-java Library Javadoc for more information.
+     * @throws LibvirtException
+     */
+    private static synchronized void setupEventListener() throws LibvirtException {
+        if (libvirtEventThread == null || !libvirtEventThread.isAlive()) {
+            // Registers a default event loop, must be called before connecting to hypervisor
+            Library.initEventLoop();
+            libvirtEventThread = new Thread(() -> {
+                while (true) {
+                    try {
+                        // This blocking call contains a loop of its own that will process events until the event loop is stopped or exception is thrown.
+                        Library.runEventLoop();
+                    } catch (LibvirtException e) {
+                        s_logger.error("LibvirtException was thrown in event loop: ", e);
+                    } catch (InterruptedException e) {
+                        s_logger.error("Libvirt event loop was interrupted: ", e);
+                    }
+                }
```

Comment:
Perhaps we could extract this block to a method (and turn the comment into a javadoc), and pass the method to the thread. What do you think, @mlsorensen?

Reply:
The comment doesn't apply to the while loop, it's just a reminder of what `Library.runEventLoop()` does. `Library.runEventLoop()` already has javadoc though so I'd be inclined to just remove the comment - the reason I added it is because it is not immediately obvious whether `Library.runEventLoop()` is just running event loop once, or calling `virEventRunDefaultImpl` repeatedly as required by the libvirt event framework.

Reply:
Makes sense. In this case, I would be inclined to remove the comment as well.